File: SngJuni/SWEA/swea11315.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

bool check(string &line) { // 특정 라인의 o 갯수를 세서 오목을 판정하는 함수
    int cnt = 0;
    for (auto c : line) {
        if (c == 'o') cnt++;
        else cnt = 0;
        if (cnt >= 5) return true;
    }
    return false;
}

int main() {
    int t;
    cin >> t;

    for (int tc = 1; tc < t + 1; tc++) {
        int n;
        cin >> n;
        vector<string> arr(n);

        for (int i = 0; i < n; i++) cin >> arr[i];

        bool flag = false;

        // 가로 검사
        for (int i = 0; i < n; i++) {
            if (check(arr[i])) flag = true;
        }

        // 세로 검사
        for (int i = 0; i < n; i++) {
            string temp;
            for (int j = 0; j < n; j++) {
                temp += arr[j][i];
            }
            if (check(temp)) flag = true;
        }

        // 우하향 대각선 (왼쪽 변에서 시작) 검사
        for (int start = 0; start < n; start++) {
            string temp;
            for (int i = start, j = 0; i < n && j < n; i++, j++) {
                temp += arr[i][j];
            }
            if (check(temp)) flag = true;
        }
        // 우하향 대각선 (윗쪽 변에서 시작) 검사
        for (int start = 1; start < n; start++) {
            string temp;
            for (int i = 0, j = start; i < n && j < n; i++, j++) {
                temp += arr[i][j];
            }
            if (check(temp)) flag = true;
        }

        // 우상향 대각선 (왼쪽 변에서 시작) 검사
        for (int start = 0; start < n; start++) {
            string temp;
            for (int i = start, j = 0; i >= 0 && j < n; i--, j++) {
                temp += arr[i][j];
            }
            if (check(temp)) flag = true;
        }
        // 우상향 대각선 (아래쪽 변에서 시작) 검사
        for (int start = 1; start < n; start++) {
            string temp;
            for (int i = n - 1, j = start; i >= 0 && j < n; i--, j++) {
                temp += arr[i][j];
            }
            if (check(temp)) flag = true;
        }

        cout << "#" << tc << " ";
        cout << (flag ? "YES" : "NO") << "\n";
    }

    return 0;
}

File: nancykim99/swea/swea11315.py
# N*N 크기의 판
# 가로, 세로, 대각선 : 5개 이상
# T : 테스트케이스
# N : 5 <= N <= 20
# o : 돌이 있는 칸
# . : 돌이 없는 칸

T = int(input())

for tc in range(1, T+1):
    N = int(input())
    arr = [list(input()) for _ in range(N)]

    flag = False

    for i in range(N):
        cnt = 0
        for j in range(N):
            if arr[i][j] == 'o':
                cnt += 1
                if cnt >= 5:
                    flag = True
                    break
            else:
                cnt = 0
        if flag:
            break
    
    if not flag: #flag가 앞에서 아직 False인 경우
        for j in range(N):
            cnt = 0
            for i in range(N):
                if arr[i][j] == 'o':
                    cnt += 1
                    if cnt >= 5:
                        flag = True
                        break
                else:
                    cnt = 0
            if flag:
                break
        
    if not flag:
        for k in range(N):
            cnt = 0
            i, j = 0, k
            while i < N and j < N:
                if arr[i][j] == 'o':
                    cnt += 1
                    if cnt >= 5:
                        flag = True
                        break
                else:
                    cnt = 0
                i += 1
                j += 1
            if flag:
                break
        
        if not flag:
            for k in range(1, N):
                cnt = 0
                i, j = k, 0
                while i < N and j < N:
                    if arr[i][j] == 'o':
                        cnt += 1
                        if cnt >= 5:
                            flag = True
                            break
                    else:
                        cnt = 0
                    i += 1
                    j += 1
                if flag:
                    break
    
    if not flag:
        for k in range(N):
            cnt = 0
            i, j = 0, k
            while i < N and j >= 0:
                if arr[i][j] == 'o':
                    cnt += 1
                    if cnt >= 5:
                        flag = True
                        break
                else:
                    cnt = 0
                i += 1
                j -= 1
            if flag:
                break
        
        if not flag:
            for k in range(1, N):
                cnt = 0
                i, j = k, N-1
                while i < N and j >= 0:
                    if arr[i][j] == 'o':
                        cnt += 1
                        if cnt >= 5:
                            flag = True
                            break
                    else:
                        cnt = 0
                    i += 1
                    j -= 1
                if flag:
                    break
    
    if flag:
        print(f'#{tc} YES')
    else:
        print(f'#{tc} NO')